### PR TITLE
Decode SourceTableColumn SQL proto.

### PR DESCRIPTION
**Pull request by markww**
This PR decodes the SourceTableColumn SQL proto. The decoding of this
proto is fairly straightforward, as it just involves getting the storage
indicated by the proto (creating it if it does not yet exist) and
wrapping the whole thing in a `StorageValue` and a `Value`.

### CODE REVIEW

File: src/ir/proto/sql/decode.cc [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  // Note: In the future, we will probably want to make a `Storage` per table
+  // and grab columns from that table using operators. For now, however,
+  // making each column an independent storage works for the MVP and is much
+  // simpler.
+  //
+  // Also, for now, we consider all storages to have primitive type. We will
+  // probably want to change that when we start handling types in a
+  // non-trivial fashion.
+  return ir::Value{value::StoredValue(
+      ir_context.GetOrCreateStorage(
```
Comment (bgogul) [inline src/ir/proto/sql/decode.cc:35]: I commented on the parent PR. Ideally, we should setup the `IRContext` with all the `Operators` and `Storage` and use it here. Adding things on the fly can be error prone. e.g., if we add more attributes to `Storage`, how do we ensure that all uses are consistent. 

With `GetOrCreateXYZ` variants, it looks like the first occurrence wherever it is wins?

File: src/ir/value.h [hunk 2]
```diff
@@ -114,6 +114,15 @@ class Value {
...
 
+  // A downcast operation. Just delegates directly to std::get on variants.
+  template<class T>
+  const T &As() const { return std::get<T>(value_); }
+
+  // A dynamic downcast operation. Just delegates directly to std::get_if on
+  // variants.
+  template<class T>
+  const T *If() const { return std::get_if<T>(&value_); }
+
```
Comment (bgogul) [inline src/ir/value.h:125]: Are these needed anymore?

File: src/ir/proto/sql/decode.cc (hunk 1 above)
Comment (markww) [reply src/ir/proto/sql/decode.cc:35]: Yes. One way to guard against the errors that you are describing might be to add a `isStaticallyIdentical` option to `Value`s, which indicates if the item is known at static time to be an identical value. Then, if we attempt to call `GetOrCreateNode` with a candidate `Node` and that `Node` is not `isStaticallyIdentical` to the value currently in the map, we can error.

For `Operator` and `Storage`, `isStaticallyIdentical` would currently just be that the names (and in the case of `Storage`, types) match. We can default it to `false` for everything else until we need this concept there.

File: src/ir/value.h (hunk 2 above)
Comment (markww) [reply src/ir/value.h:125]: I'm using `As` in some tests, either in this PR or follow-on PRs. I don't think I'm using `If` at the moment, but it seems like a useful thing to have if we have `As`. While visitors are indeed a better pattern for full visitation of a tree, sometimes it is nice to just get the single-step downcast to the specific type of value.

File: src/ir/proto/sql/decode.cc (hunk 1 above)
Comment (markww) [reply src/ir/proto/sql/decode.cc]: This is now fixed.

File: src/ir/proto/sql/decode.cc [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  // Note: In the future, we will probably want to make a `Storage` per table
+  // and grab columns from that table using operators. For now, however,
+  // making each column an independent storage works for the MVP and is much
+  // simpler.
+  //
+  // Also, for now, we consider all storages to have primitive type. We will
+  // probably want to change that when we start handling types in a
+  // non-trivial fashion.
+  return ir::Value{value::StoredValue(
+      decoder_context.GetOrCreateStorage(column_path))};
```
Comment (bgogul) [inline src/ir/proto/sql/decode.cc:35]: Don't need `ir::` prefix. Also, `{` should `(`?
```suggestion
  return Value(value::StoredValue(
      decoder_context.GetOrCreateStorage(column_path)));
```

File: src/ir/value.h (hunk 2 above)
Comment (bgogul) [reply src/ir/value.h:125]: A visitor would be ideal, but to unblock let us use `If<T>` variant. The `As<T>` variant can throw an exception and it is not obvious at use sites. With `If` variant, you will get a `nullptr`, and we can easily document use `CHECK_NOTNULL` to document expectations.

Also, add tests for `If` (in a follow up or separate PR?)

**Review by bgogul — APPROVED** [review]
Looks good assuming you get rid of the `As<T>` variant and replace uses of `As` with `If<T>` variant.

File: src/ir/value.h (hunk 2 above)
Comment (markww) [reply src/ir/value.h:125]: OK, that sounds good.

The issue with a `Visitor` here is that these are `Value`s; as such, the `Value`s that are the predecessors to this `Value` are not a tree, but a graph - the dataflow graph. Creating a `Visitor` on the `Value`s would be akin to performing a dataflow graph traversal, and I'm not sure it makes sense to do so when we want to lean on Souffle for that.

File: src/ir/proto/sql/decode.cc (hunk 3 above)
Comment (markww) [reply src/ir/proto/sql/decode.cc]: Done.

File: src/ir/value.h (hunk 2 above)
Comment (markww) [reply src/ir/value.h:125]: Using `If` instead of `As` in the test now.

**Pull request merged**